In [44]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
import numpy as np

In [16]:
patient_info = pd.read_csv("Patient_info.csv")

print("Columns:")
print(patient_info.columns.tolist())

print("\nShape:")
print(patient_info.shape)

print("\nFirst 5 rows:")
print(patient_info.head().to_string(index=False))

Columns:
['Patient_ID', 'Sex', 'Birth_year', 'Initial_measurement_date', 'Final_measurement_date', 'Number_of_days_with_measures', 'Number_of_measurements', 'Initial_biochemical_parameters_date', 'Final_biochemical_parameters_date', 'Number_of_biochemical_parameters', 'Number_of_diagnostics']

Shape:
(736, 11)

First 5 rows:
Patient_ID Sex  Birth_year Initial_measurement_date Final_measurement_date  Number_of_days_with_measures  Number_of_measurements Initial_biochemical_parameters_date Final_biochemical_parameters_date  Number_of_biochemical_parameters  Number_of_diagnostics
 LIB193263   M        1965               2020-06-09             2022-03-19                           648                   60097                                 NaN                               NaN                               NaN                    1.0
 LIB193264   F        1975               2020-06-10             2022-03-19                           326                   26786                                 

In [17]:
print("\nSex column dtype:")
print(patient_info["Sex"].dtype)

print("\nUnique Sex values:")
print(patient_info["Sex"].value_counts(dropna=False))


Sex column dtype:
object

Unique Sex values:
Sex
F    373
M    363
Name: count, dtype: int64


In [18]:
TARGET_PATIENT_ID = "LIB193263"

row = patient_info[patient_info["Patient_ID"] == TARGET_PATIENT_ID]

print("Rows found:", len(row))
print(row.to_string(index=False))

Rows found: 1
Patient_ID Sex  Birth_year Initial_measurement_date Final_measurement_date  Number_of_days_with_measures  Number_of_measurements Initial_biochemical_parameters_date Final_biochemical_parameters_date  Number_of_biochemical_parameters  Number_of_diagnostics
 LIB193263   M        1965               2020-06-09             2022-03-19                           648                   60097                                 NaN                               NaN                               NaN                    1.0


In [19]:
# Sanity check: exactly one row
if len(row) != 1:
    raise RuntimeError("Expected exactly one row for the patient")

In [20]:
# Extract original sex value
sex_value = row.iloc[0]["Sex"]

# Encode gender based on ACTUAL dataset values
# M -> 1, F -> 0
if sex_value == "M":
    gender = 1
elif sex_value == "F":
    gender = 0
else:
    raise RuntimeError(f"Unexpected Sex value: {sex_value}")

In [21]:
# Final minimal dataframe
patient_gender = pd.DataFrame({
    "Patient_ID": [TARGET_PATIENT_ID],
    "Gender": [gender]
})

# Plain-text confirmation
print("Patient ID:", TARGET_PATIENT_ID)
print("Original Sex value:", sex_value)
print("Encoded Gender:", gender)

Patient ID: LIB193263
Original Sex value: M
Encoded Gender: 1


In [22]:
patient_gender.to_csv("patient_1.csv", index=False)

In [10]:
# Read just the first 61k rows
head_chunk = pd.read_csv(
    "Glucose_measurements.csv",
    nrows=60098
)

print("Unique Patient_IDs in first 60098 rows:")
print(head_chunk["Patient_ID"].unique())


Unique Patient_IDs in first 60098 rows:
['LIB193263']


In [13]:
patient1_glucose = pd.read_csv(
    "Glucose_measurements.csv",
    nrows=60098
)

print("Shape:", patient1_glucose.shape)
print("Unique patients:", patient1_glucose["Patient_ID"].unique())


Shape: (60098, 4)
Unique patients: ['LIB193263']


In [14]:
patient1_glucose.to_csv("patient1_glucose.csv", index=False)

In [15]:
print("Missing Measurement_date:", patient1_glucose["Measurement_date"].isna().sum())
print("Missing Measurement_time:", patient1_glucose["Measurement_time"].isna().sum())
print("Missing Measurement:", patient1_glucose["Measurement"].isna().sum())


Missing Measurement_date: 0
Missing Measurement_time: 0
Missing Measurement: 0


In [24]:
cgm = pd.read_csv("patient1_glucose.csv")

print("Shape:", cgm.shape)
print("Columns:", cgm.columns.tolist())
print("\nFirst 5 rows:")
print(cgm.head().to_string(index=False))
print("\nLast 5 rows:")
print(cgm.tail().to_string(index=False))

Shape: (60098, 4)
Columns: ['Patient_ID', 'Measurement_date', 'Measurement_time', 'Measurement']

First 5 rows:
Patient_ID Measurement_date Measurement_time  Measurement
 LIB193263       2020-06-09         19:08:00           99
 LIB193263       2020-06-09         19:23:00           92
 LIB193263       2020-06-09         19:38:00           86
 LIB193263       2020-06-09         19:53:00           85
 LIB193263       2020-06-09         20:08:00           85

Last 5 rows:
Patient_ID Measurement_date Measurement_time  Measurement
 LIB193263       2022-03-19         06:32:00          124
 LIB193263       2022-03-19         06:47:00          124
 LIB193263       2022-03-19         07:02:00          118
 LIB193263       2022-03-19         07:17:00          110
 LIB193263       2022-03-19         07:32:00          105


In [25]:
print("Unique Patient_IDs:", cgm["Patient_ID"].unique())
print("Number of unique patients:", cgm["Patient_ID"].nunique())

Unique Patient_IDs: ['LIB193263']
Number of unique patients: 1


In [27]:
print("Missing values per column:")
print(cgm.isna().sum())

Missing values per column:
Patient_ID          0
Measurement_date    0
Measurement_time    0
Measurement         0
dtype: int64


In [28]:
print("Glucose stats:")
print(cgm["Measurement"].describe())

Glucose stats:
count    60098.000000
mean       177.817248
std         62.699321
min         40.000000
25%        132.000000
50%        173.000000
75%        219.000000
max        492.000000
Name: Measurement, dtype: float64


In [29]:
print("First 10 Measurement_time values:")
print(cgm["Measurement_time"].head(10).tolist())

print("\nLast 10 Measurement_time values:")
print(cgm["Measurement_time"].tail(10).tolist())

First 10 Measurement_time values:
['19:08:00', '19:23:00', '19:38:00', '19:53:00', '20:08:00', '20:23:00', '20:38:00', '20:53:00', '21:08:00', '21:23:00']

Last 10 Measurement_time values:
['05:17:00', '05:32:00', '05:47:00', '06:02:00', '06:17:00', '06:32:00', '06:47:00', '07:02:00', '07:17:00', '07:32:00']


1. Structure

60,098 rows ✔

1 patient only ✔

4 columns, all filled ✔

2. Time behavior

Exactly 15-minute sampling ✔

Continuous over ~2 years ✔

No missing times ✔

3. Glucose values

Min = 40 mg/dL ✔

Max = 492 mg/dL ✔

All values physiologically plausible ✔

No NaNs ✔

In [30]:
# Create a datetime timestamp
cgm["Timestamp"] = pd.to_datetime(
    cgm["Measurement_date"] + " " + cgm["Measurement_time"]
)

# Sort strictly by time
cgm = cgm.sort_values("Timestamp").reset_index(drop=True)

print("First timestamp:", cgm["Timestamp"].iloc[0])
print("Last timestamp:", cgm["Timestamp"].iloc[-1])


First timestamp: 2020-06-09 19:08:00
Last timestamp: 2022-03-19 07:32:00


In [33]:
cgm = cgm[["Timestamp", "Measurement"]]

print(cgm.head().to_string(index=False))
print(cgm.tail().to_string(index=False))

          Timestamp  Measurement
2020-06-09 19:08:00           99
2020-06-09 19:23:00           92
2020-06-09 19:38:00           86
2020-06-09 19:53:00           85
2020-06-09 20:08:00           85
          Timestamp  Measurement
2022-03-19 06:32:00          124
2022-03-19 06:47:00          124
2022-03-19 07:02:00          118
2022-03-19 07:17:00          110
2022-03-19 07:32:00          105


In [34]:
time_diff = cgm["Timestamp"].diff().dropna()

print("Time difference (minutes) summary:")
print(time_diff.dt.total_seconds().div(60).describe())

Time difference (minutes) summary:
count    60097.000000
mean        15.515317
std          6.993941
min          0.000000
25%         15.000000
50%         15.000000
75%         15.000000
max        555.000000
Name: Timestamp, dtype: float64


In [36]:
# Count duplicate timestamps
duplicate_timestamps = cgm.duplicated(subset=["Timestamp"]).sum()

duplicate_timestamps

np.int64(29)

In [37]:
cgm[cgm.duplicated(subset=["Timestamp"], keep=False)].head(10)

,Timestamp,Measurement
12804,2020-10-25 02:02:00,221
12805,2020-10-25 02:02:00,197
12806,2020-10-25 02:17:00,194
12807,2020-10-25 02:17:00,218
12808,2020-10-25 02:32:00,211
12809,2020-10-25 02:32:00,194
12810,2020-10-25 02:47:00,192
12811,2020-10-25 02:47:00,204
47130,2021-10-31 02:01:00,269
47131,2021-10-31 02:01:00,280


In [38]:
cgm["dt_minutes"] = cgm["Timestamp"].diff().dt.total_seconds() / 60

cgm.sort_values("dt_minutes", ascending=False).head(5)[
    ["Timestamp", "dt_minutes"]
]

,Timestamp,dt_minutes
59271,2022-03-10 09:00:00,555.0
46437,2021-10-23 18:50:00,320.0
47756,2021-11-06 18:56:00,319.0
11589,2020-10-12 00:49:00,299.0
50144,2021-12-01 18:34:00,296.0


In [45]:
# Extract glucose values as array
glucose_values = cgm["Measurement"].values.reshape(-1, 1)

In [47]:
# Decide train/val/test split (chronological)
train_size = int(len(glucose_values) * 0.7)
val_size   = int(len(glucose_values) * 0.15)
test_size  = len(glucose_values) - train_size - val_size

train_data = glucose_values[:train_size]
val_data   = glucose_values[train_size:train_size+val_size]
test_data  = glucose_values[train_size+val_size:]

In [49]:
# Initialize scaler
scaler = MinMaxScaler(feature_range=(0,1))

# Fit ONLY on training data
scaler.fit(train_data)

# Transform all sets
train_scaled = scaler.transform(train_data)
val_scaled   = scaler.transform(val_data)
test_scaled  = scaler.transform(test_data)

# Optional: store scaler for inverse-transform later

In [50]:
print("Train min/max:", train_scaled.min(), train_scaled.max())
print("Val min/max:", val_scaled.min(), val_scaled.max())
print("Test min/max:", test_scaled.min(), test_scaled.max())

Train min/max: 0.0 0.9999999999999999
Val min/max: 0.008849557522123894 0.9668141592920353
Test min/max: 0.028761061946902658 0.9867256637168141


In [51]:
# Inspect duplicate timestamps first
dup_counts = cgm.duplicated(subset=["Timestamp"], keep=False).sum()
print("Number of duplicate rows:", dup_counts)

# Aggregate duplicates by mean
cgm = cgm.groupby("Timestamp", as_index=False)["Measurement"].mean()

# Inspect again
print(cgm.head(10).to_string(index=False))

Number of duplicate rows: 58
          Timestamp  Measurement
2020-06-09 19:08:00         99.0
2020-06-09 19:23:00         92.0
2020-06-09 19:38:00         86.0
2020-06-09 19:53:00         85.0
2020-06-09 20:08:00         85.0
2020-06-09 20:23:00         87.0
2020-06-09 20:38:00         88.0
2020-06-09 20:53:00         93.0
2020-06-09 21:08:00        106.0
2020-06-09 21:23:00        134.0


In [52]:
# Calculate differences in minutes
cgm["dt_minutes"] = cgm["Timestamp"].diff().dt.total_seconds() / 60
print("Time difference summary (after duplicates removed):")
print(cgm["dt_minutes"].describe())

# Optional: plot large gaps
print(cgm[cgm["dt_minutes"] > 20][["Timestamp","dt_minutes"]].head())

Time difference summary (after duplicates removed):
count    60068.000000
mean        15.522807
std          6.987313
min          1.000000
25%         15.000000
50%         15.000000
75%         15.000000
max        555.000000
Name: dt_minutes, dtype: float64
               Timestamp  dt_minutes
211  2020-06-12 00:01:00        29.0
499  2020-06-15 00:07:00        29.0
594  2020-06-16 00:03:00        28.0
753  2020-06-17 16:00:00        30.0
1268 2020-06-23 00:40:00        29.0


In [53]:
# Choose scaled data (train + val + test concatenated for now)
all_scaled = np.concatenate([train_scaled, val_scaled, test_scaled], axis=0)

L = 24  # lookback
H = 4   # forecast horizon

X, y = [], []

for i in range(len(all_scaled) - L - H + 1):
    x_seq = all_scaled[i:i+L, 0]
    y_seq = all_scaled[i+L:i+L+H, 0]
    # Only append if no NaNs (for gaps)
    if not (np.isnan(x_seq).any() or np.isnan(y_seq).any()):
        X.append(x_seq)
        y.append(y_seq)

X = np.array(X)
y = np.array(y)

print("Windowed data shapes:", X.shape, y.shape)

Windowed data shapes: (60071, 24) (60071, 4)


In [55]:
meta = pd.read_csv("patient_1.csv")
gender = meta["Gender"].values[0]  # 0 or 1
gender_array = np.full((X.shape[0], 1), gender)

# Optional: concatenate along feature axis for LSTM
X_lstm = np.expand_dims(X, axis=2)  # shape: (windows, L, 1)
# If adding gender as feature:
X_lstm_with_gender = np.concatenate([X_lstm, np.repeat(gender_array[:, np.newaxis, :], L, axis=1)], axis=2)
print("X shape with gender feature:", X_lstm_with_gender.shape)

X shape with gender feature: (60071, 24, 2)


In [56]:
# Number of sequences
num_samples = X_lstm_with_gender.shape[0]

# Train / val / test sizes
train_size = int(num_samples * 0.7)
val_size   = int(num_samples * 0.15)
test_size  = num_samples - train_size - val_size

print("Number of sequences:")
print(f"Train: {train_size}, Val: {val_size}, Test: {test_size}")


Number of sequences:
Train: 42049, Val: 9010, Test: 9012


In [57]:
# Split X
X_train = X_lstm_with_gender[:train_size]
X_val   = X_lstm_with_gender[train_size:train_size+val_size]
X_test  = X_lstm_with_gender[train_size+val_size:]

# Split y
y_train = y[:train_size]
y_val   = y[train_size:train_size+val_size]
y_test  = y[train_size+val_size:]

# Inspect shapes
print("Shapes after split:")
print("X_train:", X_train.shape, "y_train:", y_train.shape)
print("X_val  :", X_val.shape, "y_val  :", y_val.shape)
print("X_test :", X_test.shape, "y_test :", y_test.shape)

Shapes after split:
X_train: (42049, 24, 2) y_train: (42049, 4)
X_val  : (9010, 24, 2) y_val  : (9010, 4)
X_test : (9012, 24, 2) y_test : (9012, 4)


In [58]:
# Check first input sequence
print("First X_train sequence (first 5 timesteps):")
print(X_train[0][:5, :])

print("\nCorresponding y_train sequence:")
print(y_train[0])

First X_train sequence (first 5 timesteps):
[[0.13053097 1.        ]
 [0.11504425 1.        ]
 [0.10176991 1.        ]
 [0.09955752 1.        ]
 [0.09955752 1.        ]]

Corresponding y_train sequence:
[0.33628319 0.30088496 0.27212389 0.24778761]
